### ISSUE TITLE
Separating shutdown behavior from SmalltalkCI>>test

### ISSUE THREAD

**Issue by fniephaus** [hpi-swa/smalltalkCI#242]
Currently, [SmalltalkCI>>test also exports the builds results and closes the image at the end when necessary](https://github.com/hpi-swa/smalltalkCI/blob/f54764c270b2d37289bf1efafd6db0276f6314f9/repository/SmalltalkCI-Core.package/SmalltalkCI.class/instance/test.st#L17). @dalehenrich would you object to let it return the test results instead of doing what it does now? This would allow to use SmalltalkCI's Smalltalk API in a more general way, but I guess we'd need to at least also update GsDevKit again.

**Comment by dalehenrich** [comment]
@fniephaus what specifically are you planning on changing? The semantics of #test right now is to run the test, report the results and then quit the image ...

**Comment by fniephaus** [comment]
I guess I (finally) came to the conclusion that it'd be better if #test only runs the tests. Reporting the results (currently writing a file) and quitting the image are separate steps and smalltalkCI needs to continue to provide them, but maybe in separate methods.
Instead of doing just [this](https://github.com/hpi-swa/smalltalkCI/blob/f54764c270b2d37289bf1efafd6db0276f6314f9/gemstone/run.sh#L284), we would have something like:

```Smalltalk
result := (Smalltalk at: #SmalltalkCI) test: '${config_ston}' named: '${config_smalltalk} Server (${STONE_NAME})'.
(Smalltalk at: #SmalltalkCI) reportToBash: result.
(Smalltalk at: #SmalltalkCI) shutDown: result.
```
Extending smalltalkCI's API like this would obviously allow more use cases.

**Comment by dalehenrich** [comment]
Then you are right changing the SmalltalkCI API will cause me to have to change all of the places where I depend on the old behavior ... and of course anyone else depending upon that behavior will be affected as well ...

**Comment by dalehenrich** [comment]
Perhaps you should save up all of your API changes for a couple of months so that we can go a couple of months without changing GsDevKit:)

**Comment by fniephaus** [comment]
That's why I was asking. I don't think anyone else except your and my projects depend on the API.
Since GsDevKit heavily depends on smalltalkCI, maybe it'd be better to also have the code snippets that use the API in the `gemstone/` directory? This way GsDevKit need to know less about smalltalkCI and we can maintain the code in one repo. Not sure if you'd consider this an improvement?

**Comment by dalehenrich** [comment]
Well the issue is that you have a public api and I am using the public api ... you really should not be going around changing the public API avery month or so ... I'm sure you are aware of [Semantic Versioning](http://semver.org/) and the notion of API breaking changes ... changing your API is a big thing ... and at this point in time, you only think that GsDevKit is the only project that depends upon the API, however, if you want to be a responsible developer, you have to assume that **everyone** is using the public api and go through a regular process of deprecation, removal and version changing ...

**Comment by fniephaus** [comment]
You are absolutely right, but this is not as easy when projects like GsDevKit use the latest version of smalltalkCI. If you lock it down on a specific version you would be able to decide when to upgrade to a newer version of smalltalkCI. And I would be able to release 2.0 with breaking API changes in the meantime. This of course gets complicated again when we need to fix a bug in an older version and the migration to the latest version is not an option.
I could also start decoding versions in method names, but then we are bloating smalltalkCI with even more compatibility code.
Do you have a good idea how we could enhance smalltalkCI without impacting API users too much? I know you are busy too, so I don't want to cause any additional work for you every time I want to change something. This is why I started this thread in the first place :)

**Comment by fniephaus** [comment]
Oh right, GsDevKit is a dependency of smalltalkCI, and it doesn't make much sense to clone a specific version of that. smalltalkCI could check out an older version of itself when necessary but that also seems to be a bad idea...

**Comment by dalehenrich** [comment]
You need to have a process where you deprecate features and a policy for when the deprecated api is changed ... along with a regular release cycle ... I believe that 1.0.0 was released a year ago and AFAICT there have been a number of API breaking changes made since then without a new release or any deprecation ...

Once you have a process that includes deprecation and regular releases based on Semantic versioning, I will change the way that I use SmalltalkCI --- I will use tags to control what I use for GsDevKit_home and will pay special attention when a new major version of SmalltalkCI comes out because that means that there are API breaking changes ---

**Comment by dalehenrich** [comment]
Regarding process I would look around at what other github projects do to manage there versions ... 

I would think that tags along the master branch would work fine and version specific branches for bugfixes to older versions would be fine ... the key is to deprecate, release, implement API breaking changes and increment major version number ...

If you start this process, I can then alter the way that I checkout SmalltalkCI to protect myself from API breaking changes ...

**Comment by fniephaus** [comment]
I've marked a commit that worked fine after the last big changes in October as [v2.0.0](https://github.com/hpi-swa/smalltalkCI/releases/tag/v2.0.0). When https://github.com/hpi-swa/smalltalkCI/pull/243 is merged, this will be v2.1.0. And from now on, whenever we merge something into master, it will get an appropriate version tag.
As soon as you are ok with me changing the behavior of `SmalltalkCI>>test:`, we'll have `v3.0.0`. But it's ok if you want to wait with that for a couple of weeks.

**Comment by fniephaus** [comment]
When this is done, `load` and `test` should return something more meaningful, e.g.: `true` if successful, otherwise `false`.
